doc/../rust.section.md: fix incorrect header depths (#187919)
NixOS/nixpkgs: doc/languages-frameworks/rust.section.md

}
```

> ## `buildRustCrate`: Compiling Rust crates using Nix instead of Cargo {#compiling-rust-crates-using-nix-instead-of-cargo}

Reviewer: ```suggestion
## `buildRustCrate`: Compiling Rust crates using Nix instead of Cargo {#compiling-rust-crates-using-nix-instead-of-cargo}

Note that crate2nix needs reimplements build logic from cargo. This method can fail for crates that rely on features not implemented yet i.e. workspaces.
```

Reviewer: Correct me if I am wrong, but this is my last information I had about it.

Author: I think this amounts to "XYZ won't work if it has bugs", which you could say about any piece of software, including Nix.  Of the four restrictions on that page, one is believed-to-be-resolved and one is actually a Nix limitation:

* A crate will only have access to its own source directory during build time

How will we remember to delete this paragraph when the three remaining outstanding issues are resolved?

I think anybody considering using `crate2nix` is going to read the `crate2nix` documentation anyways.

Another alternative would be to simply remove all the `crate2nix` references since it isn't part of nixpkgs.  But `buildRustCrate` is.

Author: I resolved this by simply eliminating all references to `crate2nix` other than linking to its docs.  `buildRustCrate` is the only thing we should be documenting since it is the only thing that is part of nixpkgs.

### Simple operation {#simple-operation}

When run, `cargo build` produces a file called `Cargo.lock`,
containing pinned versions of all dependencies. Nixpkgs contains a

> tool called `carnix` (`nix-env -iA nixos.carnix`), which can be used

Reviewer: ```suggestion
tool called `crate2Nix` (`nix-shell -p crate2nix`), which can be used
```

to turn a `Cargo.lock` into a Nix expression.

That Nix expression calls `rustc` directly (hence bypassing Cargo),
and can be used to compile a crate and all its dependencies. Here is
an example for a minimal `hello` crate:

```ShellSession
$ cargo new hello
$ cd hello
$ cargo build
     Compiling hello v0.1.0 (file:///tmp/hello)
     Finished dev [unoptimized + debuginfo] target(s) in 0.20 secs

> $ carnix -o hello.nix --src ./. Cargo.lock --standalone

Reviewer: crate2nix generate?

$ nix-build hello.nix -A hello_0_1_0
```

Now, the file produced by the call to `carnix`, called `hello.nix`, looks like:

```nix

> # Generated by carnix 0.6.5: carnix -o hello.nix --src ./. Cargo.lock --standalone

Reviewer: needs up-date.

{ stdenv, buildRustCrate, fetchgit }:
let kernel = stdenv.buildPlatform.parsed.kernel.name;
    # ... (content skipped)
in
rec {
  hello = f: hello_0_1_0 { features = hello_0_1_0_features { hello_0_1_0 = f; }; };
  hello_0_1_0_ = { dependencies?[], buildDependencies?[], features?[] }: buildRustCrate {
    crateName = "hello";
    version = "0.1.0";
    authors = [ "pe@pijul.org <pe@pijul.org>" ];
    src = ./.;
    inherit dependencies buildDependencies features;
  };
  hello_0_1_0 = { features?(hello_0_1_0_features {}) }: hello_0_1_0_ {};
  hello_0_1_0_features = f: updateFeatures f (rec {
        hello_0_1_0.default = (f.hello_0_1_0.default or true);
    }) [ ];
}
```

In particular, note that the argument given as `--src` is copied
verbatim to the source. If we look at a more complicated
dependencies, for instance by adding a single line `libc="*"` to our
`Cargo.toml`, we first need to run `cargo build` to update the

> `Cargo.lock`. Then, `carnix` needs to be run again, and produces the

Reviewer: ```suggestion
`Cargo.lock`. Then, `crate2nix` needs to be run again, and produces the
```

following nix file:

```nix

> # Generated by carnix 0.6.5: carnix -o hello.nix --src ./. Cargo.lock --standalone

Reviewer: Needs update

  (hello {}).override { release = false; };
  ```

- Whether to print the commands sent to `rustc` when building
  (equivalent to `--verbose` in cargo:

  ```nix
  (hello {}).override { verbose = false; };
  ```

- Extra arguments to be passed to `rustc`:

  ```nix
  (hello {}).override { extraRustcOpts = "-Z debuginfo=2"; };
  ```

- Phases, just like in any other derivation, can be specified using
  the following attributes: `preUnpack`, `postUnpack`, `prePatch`,
  `patches`, `postPatch`, `preConfigure` (in the case of a Rust crate,
  this is run before calling the "build" script), `postConfigure`
  (after the "build" script),`preBuild`, `postBuild`, `preInstall` and
  `postInstall`. As an example, here is how to create a new module
  before running the build script:

  ```nix
  (hello {}).override {
    preConfigure = ''
       echo "pub const PATH=\"${hi.out}\";" >> src/path.rs"
    '';
  };
  ```

### Features {#features}

One can also supply features switches. For example, if we want to
compile `diesel_cli` only with the `postgres` feature, and no default
features, we would write:

```nix
(callPackage ./diesel.nix {}).diesel {
  default = false;
  postgres = true;
}
```

> Where `diesel.nix` is the file generated by Carnix, as explained above.

Reviewer: ```suggestion
Where `diesel.nix` is the file generated by crate2nix, as explained above.
```

$ cargo test
```

> #### Using community maintained Rust toolchains {#using-community-maintained-rust-toolchains}

Reviewer 2: This is not part of "Setting up `nix-shell`", it should belong to its own section where "Using Rust nightly with `nix-shell`" and "Using Rust nightly in a derivation with `buildRustPackage`" are its subsections

Author: Thanks for catching that; I rebased incorrectly.
